Combined PR #1512 and #1598

sherylll opened the pull request:
Hi,

please check out the reworked version of PR #1512. I merged the two branches and changes are: 
- combined `UA_Client_run_iterate` and `UA_Client_runAsync`, all the features mentioned in the latter should still work (Without canceling timed out requests feature and InactivityCheck is not fully tested)
- reworked the unit test `check_client_subscription` and removed the assertions on the number of nontifications since with the non-blocking implementation this may vary each time, I simply checked if any notification is received.

StalderT commented on src/client/ua_client_highlevel.c:
```diff
@@ -54,173 +56,171 @@ UA_Client_NamespaceGetIndex(UA_Client *client, UA_String *namespaceUri,
-    UA_BrowseRequest_deleteMembers(&bReq);
-    UA_BrowseResponse_deleteMembers(&bResp);
-    return retval;
+UA_StatusCode UA_Client_forEachChildNodeCall(UA_Client *client,
```
To reduce the PR size can you revert the identation for all function in this file?

StalderT commented on src/client/ua_client_connect.c:
```diff
@@ -26,10 +26,11 @@
 void
-setClientState(UA_Client *client, UA_ClientState state) {
-    if(client->state != state) {
+setClientState(UA_Client *client, UA_ClientState state)
```
to reduce the PR size can you revert the indentation in this file

StalderT commented on src/client/ua_client.c:
```diff
@@ -329,7 +338,7 @@ processAsyncResponse(UA_Client *client, UA_UInt32 requestId, const UA_NodeId *re
 
     /* Call the callback */
-    ac->callback(client, ac->userdata, requestId, response, ac->responseType);
+    ac->callback(client, ac->userdata, requestId, response);
```
Why do you have remove ```ac->responseType``` ?

sherylll replied:
Even though it's defined the `responseType` is never actually needed. In 'ua_client_highlevel_async.h' I wrapped the type inside the high-level functions

StalderT replied:
Ok. Thank you for your explanation.

StalderT commented on examples/server_repeated_job.c:
```diff
@@ -4,9 +4,13 @@
 #include <signal.h>
 #include "open62541.h"
 
+UA_UInt64 callbackID;
```
Can you remove changes in server part?

StalderT commented on plugins/ua_network_tcp.c:
```diff
@@ -1,11 +1,11 @@
+ * See http://creativecommons.org/publicdomain/zero/1.0/ for more information.
  *
- *    Copyright 2016-2017 (c) Fraunhofer IOSB (Author: Julius Pfrommer)
+ *    Copyright 2016-2017 (c) Julius Pfrommer, Fraunhofer IOSB
```
must be Copyright 2016-2017 (c) Fraunhofer IOSB (Author: Julius Pfrommer)

StalderT commented on include/ua_client_highlevel.h:
```diff
@@ -562,6 +562,17 @@ UA_Client_addMethodNode(UA_Client *client, const UA_NodeId requestedNewNodeId,
                                &UA_TYPES[UA_TYPES_METHODATTRIBUTES], outNewNodeId);
 }
 
+/**
```
Is this comment still relevant?

StalderT commented on src/client/ua_client_worker.c:
```diff
@@ -0,0 +1,148 @@
+ * Stop: Stop workers, finish all callbacks, stop the network layer,
+ *       clean up */
+
+UA_StatusCode UA_Client_run_iterate(UA_Client *client, UA_Boolean *timedOut) {
```
Can you explain to me what is the timeOut variable?

sherylll replied:
I followed the example of `UA_Client_runAsync(UA_Client *client, UA_UInt16 timeout)` but then I realized it's bit difficult to time this function exactly (in unit testing) so I thought the user may use `timedOut` to end it manually, similarly as in the case of `UA_Server_run_iterate` where a global variable 'running' can be used to turn the function on and off.
But you are right, in the examples available this `timedOut` is superfluous.

sherylll replied:
Yes I think we can do that. 

The main difference is the use of `UA_Connection_receiveChunksNonBlocking`, which removes the while loop as compared to `UA_Connection_receiveChunksBlocking`

StalderT replied:
Ok thanks

StalderT commented on src/client/ua_client_worker.c:
```diff
@@ -0,0 +1,148 @@
+    if(*timedOut)
+        return UA_STATUSCODE_GOODNONCRITICALTIMEOUT;
+
+    if (client->config.stateCallback)
```
Must be removed. Called by setClientState when a state change.

sherylll replied:
OK

StalderT commented:
Hi @sherylll 

I made some modifications. Now tests/client/check_client_async.c don't pass. Can you check? You have acces right to directly write to StalderT:checkconnectivity2 branch.

To not pollute open62541 I propose to close https://github.com/open62541/open62541/pull/1598 And continue discussing in this thread. We will reopen the PR when it is ready. Ok?

sherylll commented:
Hi,
I had a little problem pushing to your repository. I accidentally created another branch `StalderT/checkconnectivity2` (should be only `checkconnectivity2`). Then I created a new branch `wrong_branch` locally, and now I want to rebase it onto checkconnectivity2, but git wouldn't let mer do so and it says 
> fatal: Needed a single revision
Does not point to a valid commit: checkconnectivity2

What should I do now?
I think it's okay to close #1598 temporarily  (now that I'm already polluting the history :()

StalderT commented:
can you go to irc https://webchat.freenode.net/?channels=%23open62541

StalderT commented:
yes fakeSleep fix the tests

StalderT commented:
else make a PR for checkconnectivity2

StalderT commented:
travis all green 👍 

can you check appveyor error?

sherylll commented:
Hi I can't see the appveyor build have you pushed anything after this 52df9f9?

StalderT commented:
> Hi I can't see the appveyor build have you pushed anything after this 52df9f9?

https://github.com/StalderT/open62541/commits/checkconnectivity2
https://github.com/StalderT/open62541/commit/b7d77a30767716070e0b54994abc02dc51c96a37 is the last commit

Appveyor build for this commit : https://ci.appveyor.com/project/StalderT/open62541/build/651/job/aev981ddjg8bbeen#L11782

Same error with last commit of your PR : https://ci.appveyor.com/project/StalderT/open62541/build/638/job/ki43j2ekle3iqq1j#L11721

I propose the following roadmap : 
1) Fix appveyor (windows build)
2) Align ```ua_client_connect_async.c``` with ```ua_client_connect.c``` (particularly the changes made by https://github.com/open62541/open62541/pull/1686, check the rest)
3) Add more tests (encryption, connect error, timeout, etc...)
4) Simplify ```UA_Client_run_iterate```
5) cosmetic fix (indentation, etc...)

Do you agree ?

I still have some questions : 

1) Here https://github.com/StalderT/open62541/blob/b7d77a30767716070e0b54994abc02dc51c96a37/examples/client_async.c#L34 you change ```connected``` can we not change internally the state of this variable?

2) Can you explain in detail how ```UA_Client_connect_async``` works?

Some changes to make :

1) https://github.com/StalderT/open62541/blob/b7d77a30767716070e0b54994abc02dc51c96a37/src/client/ua_client_connect_async.c#L495 must be move at the top of the function

2) https://github.com/StalderT/open62541/blob/b7d77a30767716070e0b54994abc02dc51c96a37/src/client/ua_client_connect_async.c#L517 miss a goto cleanup

sherylll commented:
I fixed several issues you mentioned and added detailed comments to async connect. Basically the requests and responses for the connection are chained by callbacks and processed in UA_Client_run_iterate(), based on the idea that, for example, only after an OPN response is received, can the next request (session) be sent.
With regards to appveyor error, I am also very confused... Last time I simply started the build again and it was all green.
And yes, I will take a look at open62541#1686.

StalderT commented:
Tanks for the fix. I thinks there is a poblem with appveyor. Sometime is green bu test fails : https://ci.appveyor.com/project/StalderT/open62541/build/655/job/smr9wwux5ph8n24u#L11981

Then there is a problem with windows connect async. Can you check ?

Last commit https://github.com/StalderT/open62541/commit/5297bf12f2c405c3122f7929d981ec6cb8630418

StalderT commented:
I restarted travis build, all green.

An error is still present in appveyor : https://ci.appveyor.com/project/StalderT/open62541/build/659/job/9rt7ia3h5jidld60#L11855

```
  23: 0%: Checks: 1, Failures: 1, Errors: 0
  23: C:\projects\open62541\tests\client\check_client_async_connect.c:101:F:Client Connect Async:Client_connect_async:0: Assertion 'connected == 1' failed: connected == 0, 1 == 1
```

sherylll commented:
Hi,

except for the error under Windows build, I think this PR is ready to be reopened. However AppVeyor seem to have some problems https://ci.appveyor.com/project/StalderT/open62541/build/job/776iqb2lgjigag06

Any idea?

ccvca commented:
> Any idea?

Fixed by: 63ba1c0f8ba50f1e3beb990fcb2e9c08cee22b0e

StalderT commented:
In the morning. I will merge master.

I will also make some new comments about this PR.
